frontend/src/modules/machinery/components/source/GoogleTranslation.tsx

```diff
@@ -1,5 +1,189 @@
+        //
+        // When spaces are added, GTA is able to translate the input, however that introduces additional
+        // complexity related to restoring the original spacing between a placeable and its surrounding words/tokens.
+        // e.g.
+        //   Original string: On the nature of \n${something}  # (only ${something} is a placeable here)
+        //   ^ Notice the lack of trailing spaces
+        //   GTA input string: On the nature of \n <placeable hash> $
+        //   ^ the pre-processing function added the trailing spaces. $ is the end marker of the string.
+        //   GTA output string [locale: pl]: O naturze \n <placeable hash>$
+        //   ^ GTA returns only the left trailing space. $ is the end marker of the string.
+        //   Final string visible in the Translate.Next UI: O naturze \n ${something}$
+        // Also, the direction of a script in a locale (LTR, RTL) is a factor that may change the order
+        // of the placeables in a string and make restoring the surrounding spaces harder.
+
+        newText +=
+            text.substring(pos, placeableOccurrence.index) +
+            ' ' +
+            getPlaceableHash(placeable) +
+            ' ';
+        pos = placeableOccurrence.index + placeable.length;
+        placeableOccurrence = placeablesRegex.exec(text);
+    }
+
+    if (pos < text.length - 1) {
+        newText += text.substring(pos);
+    }
+
+    return newText;
+}
+
+/**
+ * Google Translate API supports two input formats:
+ * * text
+ * * html
+ * Depending on the selected format type, the API applies and returns format specific transformations,
+ * e.g. replaces some reserved characters with HTML entities.
+ */
+export function GetGoogleTranslateInputFormat(text: string): 'text' | 'html' {
+    // The regexes are borrowed from the corresponding placeable rules.
+    const htmlEntities: RegExp = /(&(([a-zA-Z][a-zA-Z0-9.-]*)|([#](\d{1,5}|x[a-fA-F0-9]{1,5})+));)/;
+    const htmlTags: RegExp = /(<[\w.:]+(\s([\w.:-]+=((".*?")|('.*?')))?)*\/?>|<\/[\w.]+>)/;
+
+    if (text.search(htmlEntities) !== -1 || text.search(htmlTags) !== -1) {
+        return 'html';
+    }
+
+    return 'text';
+}
+
+/**
+ *  Process the response from the Google Translate API and replace the placeable hashes with their original values.
+ *  Validate the translation and throw an error when a placeable is missing.
+ */
+export function GetGoogleTranslateResponseText(
+    response: any,
+    placeables: Array<string>,
+): string | null {
+    if (!response.translation) {
+        throw new Error('No translation in response');
+    }
+
+    const checkAndReplace = (acc, placeable): string => {
+        let placeableHash = getPlaceableHash(placeable);
+
+        if (response.translation.indexOf(placeableHash) == -1) {
+            throw new Error(
+                `Google Translate API removed the placeable: ${placeable}`,
+            );
+        }
+        return acc.replace(placeableHash, placeable);
```

Comment:
If the same placeable shows up more than once in a message, will it also show up multiple times in `placeables`? This `replace()` call will only replace the first instance.

Reply:
That's something I've missed, and I'll add a test case for that too. Thanks!